### ISSUE TITLE
astropy.io.fits.append ignores header keyword

### ISSUE THREAD

**Issue by sbailey** [astropy/astropy#4167]
astropy.io.fits.append ignores the header keyword:

```
from astropy.io import fits
import numpy as np

outfile = 'blat.fits'
data = np.arange(10)
fits.writeto(outfile, data, clobber=True)
hdr = fits.Header()
hdr['EXTNAME'] = 'BLAT'
hdr['BLAT'] = 'FOO'
fits.append(outfile, data, header=hdr)

fx = fits.open(outfile)
fx.info()
fx[1].header  #- missing EXTNAME and BLAT
```

Creating an HDUList ahead of time and writing out everything at once does work:

```
from astropy.io import fits
import numpy as np

outfile = 'blat.fits'
data = np.arange(10)

hx = fits.HDUList()
hx.append(fits.PrimaryHDU(data))
hdr = fits.Header()
hdr['EXTNAME'] = 'BLAT'
hdr['BLAT'] = 'FOO'
hx.append(fits.ImageHDU(data, header=hdr))

hx.writeto(outfile, clobber=True)
fx = fits.open(outfile)
fx.info()
fx[1].header
```

However, I don't always have all my HDUs ready at once and it is indeed convenient to just be able to append a new HDU to a pre-existing file with a single command.

astropy 1.0.4 with python 2.7.10 with anaconda on a mac

**Comment by embray** [comment]
Confirmed--I'm actually surprised this has never been caught before (note: the convenience functions work rather differently than using the object oriented API--fits.append() has to do some guessing as to the most appropriate thing to write, whereas using the OO API is much more explicit, which is one reason I usually recommend it).

Still surprising nonetheless.  For the benefit of myself later, or anyone else who wants to fix this the problem occurs at this line: https://github.com/astropy/astropy/blob/master/astropy/io/fits/convenience.py#L772  (or two lines up for the case of tables).

I've marked this as "intermediate" for now though, since I'm not 100% sure what other problems are lurking.  I think (believe it or not) this may also be partly a symptom of the same issue that underlies #4063 .

**Comment by MSeifert04** [comment]
I looked into it and like @embray said the problem is that it cannot find the right subclass. Actually for the ``fits.append`` convenience method to work in this case properly you have to define that it is an image xtension.

```
from astropy.io import fits
import numpy as np

outfile = 'blat.fits'
data = np.arange(10)
fits.writeto(outfile, data, clobber=True)
hdr = fits.Header()
hdr['XTENSION'] = 'IMAGE' #has to be defined as first element of the header 
#otherwise it does not recognize it as ImageHDU
hdr['EXTNAME'] = 'BLAT'
hdr['BLAT'] = 'FOO'
fits.append(outfile, data, header=hdr)

fx = fits.open(outfile)
fx.info()
fx[1].header #now works
```
But that is more than unsatifactory I guess.

**Comment by embray** [comment]
Another way you can do it is to not use the "convenience function", which is what's broken in this case, and use the (often more reliable) OO interface:
```python
new_hdu = fits.ImageHDU(data, header)  # Explicitly makes an image HDU

# Now open the file in append mode and append the extension:
with fits.open('blah.fits', mode='append') as hdu_list:
    hdu_list.append(new_hdu)
```

This is essentially what `fits.append` is doing too, but more explicit--it doesn't have to do any guessing as to what the extension type should be.

**Comment by embray** [comment]
(That said what @sbailey reported is still a bug and ought to be fixed.)

**Comment by anchitjain1234** [comment]
@embray in https://github.com/astropy/astropy/blob/master/astropy/io/fits/convenience.py#L772 if we also pass header to ``ImageHDU`` like ``hdu = ImageHDU(data=data,header=header)`` than ``fx[1].header`` works.
Same is true for Tables.
Would this resolve the issue?

**Comment by embray** [comment]
@anchitjain1234 I think that would be sufficient for at least basic cases like the one @sbailey reported in the original issue.  I think there are still unresolved problems with this interface, simply insofar as how it's a little implicit.  But your suggestion would be a good place to start nonetheless.

**Comment by saimn** [comment]
I cannot reproduce, I think it was fixed by #6058.
